src/nupic/ntypes/ArrayBase.cpp

```
@@ -233,176 +254,232 @@ void ArrayBase::setCount(size_t count) {
  */
 NTA_BasicType ArrayBase::getType() const { return type_; };
 
+/**
+ * Return true if a buffer has been allocated.
+ */
+bool ArrayBase::has_buffer() const { return (buffer_.get() != nullptr); }
+
 /**
  * Convert the buffer contents of the current ArrayBase into
- * the type of the incoming ArrayBase type. Applying an offset if specified.
+ * the type of the incoming ArrayBase 'a'. Applying an offset if specified.
+ * This may be called multiple times to set values of different offsets.
  * If there is not enough room in the destination buffer a new one is created.
+ * After allocating the buffer, zero it to clear zero values (if converting
+ * from Sparse to Dense).
+ *
  * For Fan-In condition, be sure there is enough space in the buffer before
- * the first conversion to avoid loosing data during re-allocation.
+ * the first conversion to avoid loosing data during re-allocation. Then do
+ * them in order so that the largest index is last.
+ *
+ * Be careful when using this with destination of SDR...it will remove
+ * dimensions if buffer is  not big enough.
```

can we rather throw an err in that case?

---

this ^